### DISCUSSION TITLE
postCSS issues even if postcss@^8 is already added

### DISCUSSION THREAD

**Discussion by m4th3w561** [storybookjs/storybook#19248] [Help] [answered]
So I was working with storybook the previous day and everything was working even when I fired storybook to my local server. But then when I got back to work on storybook the following day, it just started to not work and gave me these errors. 
```
ERR! TypeError [ERR_INVALID_ARG_TYPE]: The "id" argument must be of type string. Received an instance of Object
ERR!     at new NodeError (node:internal/errors:377:5)
ERR!     at validateString (node:internal/validators:119:11)
ERR!     at Module.require (node:internal/modules/cjs/loader:998:3)
ERR!     at require (node:internal/modules/cjs/helpers:102:18)
ERR!     at interopRequireDefault (...\node_modules\@storybook\core-common\dist\cjs\presets.js:178:16)
ERR!     at getContent (...\node_modules\@storybook\core-common\dist\cjs\presets.js:194:10)
ERR!     at loadPreset (...\node_modules\@storybook\core-common\dist\cjs\presets.js:203:20)
ERR!     at ...\node_modules\@storybook\core-common\dist\cjs\presets.js:256:18
ERR!     at Array.reduce (<anonymous>)
ERR!     at loadPresets (...\node_modules\@storybook\core-common\dist\cjs\presets.js:255:18)
ERR!  TypeError [ERR_INVALID_ARG_TYPE]: The "id" argument must be of type string. Received an instance of Object
ERR!     at new NodeError (node:internal/errors:377:5)
ERR!     at validateString (node:internal/validators:119:11)
ERR!     at Module.require (node:internal/modules/cjs/loader:998:3)
ERR!     at require (node:internal/modules/cjs/helpers:102:18)
ERR!     at interopRequireDefault (...\node_modules\@storybook\core-common\dist\cjs\presets.js:178:16)
ERR!     at getContent (...\node_modules\@storybook\core-common\dist\cjs\presets.js:194:10)
ERR!     at loadPreset (...\node_modules\@storybook\core-common\dist\cjs\presets.js:203:20)
ERR!     at ...\node_modules\@storybook\core-common\dist\cjs\presets.js:256:18
ERR!     at Array.reduce (<anonymous>)
ERR!     at loadPresets (...\node_modules\@storybook\core-common\dist\cjs\presets.js:255:18) {
ERR!   code: 'ERR_INVALID_ARG_TYPE'
ERR! }
ERR! TypeError: The "id" argument must be of type string. Received an instance of Object
ERR!     at new NodeError (node:internal/errors:377:5)
ERR!     at validateString (node:internal/validators:119:11)
ERR!     at Module.require (node:internal/modules/cjs/loader:998:3)
ERR!     at require (node:internal/modules/cjs/helpers:102:18)
ERR!     at interopRequireDefault (...\node_modules\@storybook\core-common\dist\cjs\presets.js:178:16)
ERR!     at getContent (...\node_modules\@storybook\core-common\dist\cjs\presets.js:194:10)
ERR!     at loadPreset (...\node_modules\@storybook\core-common\dist\cjs\presets.js:203:20)
ERR!     at ...\node_modules\@storybook\core-common\dist\cjs\presets.js:256:18
ERR!     at Array.reduce (<anonymous>)
ERR!     at loadPresets (...\node_modules\@storybook\core-common\dist\cjs\presets.js:255:18)
ERR!     at loadPreset (...\node_modules\@storybook\core-common\dist\cjs\presets.js:225:81)
ERR!     at ...\node_modules\@storybook\core-common\dist\cjs\presets.js:256:18
ERR!     at Array.reduce (<anonymous>)
ERR!     at loadPresets (...\node_modules\@storybook\core-common\dist\cjs\presets.js:255:18)
ERR!     at getPresets (...\node_modules\@storybook\core-common\dist\cjs\presets.js:313:23)
ERR!     at loadAllPresets (...\node_modules\@storybook\core-common\dist\cjs\presets.js:365:10)
ERR!     at buildDev (...\node_modules\@storybook\core-server\dist\cjs\build-dev.js:209:50)
ERR!     at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
--------------------------------------------------------------------------------------------------
ERR!  TypeError: The "id" argument must be of type string. Received an instance of Object
ERR!     at new NodeError (node:internal/errors:377:5)
ERR!     at validateString (node:internal/validators:119:11)
ERR!     at Module.require (node:internal/modules/cjs/loader:998:3)
ERR!     at require (node:internal/modules/cjs/helpers:102:18)
ERR!     at interopRequireDefault (...\node_modules\@storybook\core-common\dist\cjs\presets.js:178:16)
ERR!     at getContent (...\node_modules\@storybook\core-common\dist\cjs\presets.js:194:10)
ERR!     at loadPreset (...\node_modules\@storybook\core-common\dist\cjs\presets.js:203:20)
ERR!     at ...\node_modules\@storybook\core-common\dist\cjs\presets.js:256:18
ERR!     at Array.reduce (<anonymous>)
ERR!     at loadPresets (...\node_modules\@storybook\core-common\dist\cjs\presets.js:255:18)
ERR!     at loadPreset (...\node_modules\@storybook\core-common\dist\cjs\presets.js:225:81)
ERR!     at ...\node_modules\@storybook\core-common\dist\cjs\presets.js:256:18
ERR!     at Array.reduce (<anonymous>)
ERR!     at loadPresets (...\node_modules\@storybook\core-common\dist\cjs\presets.js:255:18)
ERR!     at getPresets (...\node_modules\@storybook\core-common\dist\cjs\presets.js:313:23)
ERR!     at loadAllPresets (...\node_modules\@storybook\core-common\dist\cjs\presets.js:365:10)
ERR!     at buildDev (...\node_modules\@storybook\core-server\dist\cjs\build-dev.js:209:50)
ERR!     at process.processTicksAndRejections (node:internal/process/task_queues:95:5) {
ERR!   code: 'ERR_INVALID_ARG_TYPE'
ERR! }
```

I am using react, next.js, tailwind, and typescript in this project.
I had some issues before too but I think I got to work around those by adding some addons to storybook (some of these might be the reason that is causing these errors?) 
Here are my addons
```
addons: [
    '@storybook/addon-links',
    "@storybook/addon-essentials",
    "@storybook/addon-interactions",
    "@storybook/addon-a11y",
    "storybook-dark-mode",
    {
      name: "@storybook/addon-postcss",
      options: {
        postcssLoaderOptions: {
          implementation: require('postcss'),
        },
      },
    },
    "storybook-tailwind-dark-mode",
    "storybook-addon-next",
    "storybook-addon-next-router",
    "storybook-addon-mock/register",
]
```

(with the post CSS it was having trouble with posttCSS8 so I remove this package and got the postCSS7)

I am not sure what makes storybook work the previous day but not work the next day, without changing anything. 

I'd appreciate it if someone can point me where I could go from here, thanks!

**Comment by m4th3w561** [comment]
> So I was working with storybook the previous day and everything was working even when I fired storybook to my local server. But then when I got back to work on storybook the following day, it just started to not work and gave me these errors.
> 
> ```
> ERR! TypeError [ERR_INVALID_ARG_TYPE]: The "id" argument must be of type string. Received an instance of Object
> ERR!     at new NodeError (node:internal/errors:377:5)
> ERR!     at validateString (node:internal/validators:119:11)
> ERR!     at Module.require (node:internal/modules/cjs/loader:998:3)
> ERR!     at require (node:internal/modules/cjs/helpers:102:18)
> ERR!     at interopRequireDefault (...\node_modules\@storybook\core-common\dist\cjs\presets.js:178:16)
> ERR!     at getContent (...\node_modules\@storybook\core-common\dist\cjs\presets.js:194:10)
> ERR!     at loadPreset (...\node_modules\@storybook\core-common\dist\cjs\presets.js:203:20)
> ERR!     at ...\node_modules\@storybook\core-common\dist\cjs\presets.js:256:18
> ERR!     at Array.reduce (<anonymous>)
> ERR!     at loadPresets (...\node_modules\@storybook\core-common\dist\cjs\presets.js:255:18)
> ERR!  TypeError [ERR_INVALID_ARG_TYPE]: The "id" argument must be of type string. Received an instance of Object
> ERR!     at new NodeError (node:internal/errors:377:5)
> ERR!     at validateString (node:internal/validators:119:11)
> ERR!     at Module.require (node:internal/modules/cjs/loader:998:3)
> ERR!     at require (node:internal/modules/cjs/helpers:102:18)
> ERR!     at interopRequireDefault (...\node_modules\@storybook\core-common\dist\cjs\presets.js:178:16)
> ERR!     at getContent (...\node_modules\@storybook\core-common\dist\cjs\presets.js:194:10)
> ERR!     at loadPreset (...\node_modules\@storybook\core-common\dist\cjs\presets.js:203:20)
> ERR!     at ...\node_modules\@storybook\core-common\dist\cjs\presets.js:256:18
> ERR!     at Array.reduce (<anonymous>)
> ERR!     at loadPresets (...\node_modules\@storybook\core-common\dist\cjs\presets.js:255:18) {
> ERR!   code: 'ERR_INVALID_ARG_TYPE'
> ERR! }
> ERR! TypeError: The "id" argument must be of type string. Received an instance of Object
> ERR!     at new NodeError (node:internal/errors:377:5)
> ERR!     at validateString (node:internal/validators:119:11)
> ERR!     at Module.require (node:internal/modules/cjs/loader:998:3)
> ERR!     at require (node:internal/modules/cjs/helpers:102:18)
> ERR!     at interopRequireDefault (...\node_modules\@storybook\core-common\dist\cjs\presets.js:178:16)
> ERR!     at getContent (...\node_modules\@storybook\core-common\dist\cjs\presets.js:194:10)
> ERR!     at loadPreset (...\node_modules\@storybook\core-common\dist\cjs\presets.js:203:20)
> ERR!     at ...\node_modules\@storybook\core-common\dist\cjs\presets.js:256:18
> ERR!     at Array.reduce (<anonymous>)
> ERR!     at loadPresets (...\node_modules\@storybook\core-common\dist\cjs\presets.js:255:18)
> ERR!     at loadPreset (...\node_modules\@storybook\core-common\dist\cjs\presets.js:225:81)
> ERR!     at ...\node_modules\@storybook\core-common\dist\cjs\presets.js:256:18
> ERR!     at Array.reduce (<anonymous>)
> ERR!     at loadPresets (...\node_modules\@storybook\core-common\dist\cjs\presets.js:255:18)
> ERR!     at getPresets (...\node_modules\@storybook\core-common\dist\cjs\presets.js:313:23)
> ERR!     at loadAllPresets (...\node_modules\@storybook\core-common\dist\cjs\presets.js:365:10)
> ERR!     at buildDev (...\node_modules\@storybook\core-server\dist\cjs\build-dev.js:209:50)
> ERR!     at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
> --------------------------------------------------------------------------------------------------
> ERR!  TypeError: The "id" argument must be of type string. Received an instance of Object
> ERR!     at new NodeError (node:internal/errors:377:5)
> ERR!     at validateString (node:internal/validators:119:11)
> ERR!     at Module.require (node:internal/modules/cjs/loader:998:3)
> ERR!     at require (node:internal/modules/cjs/helpers:102:18)
> ERR!     at interopRequireDefault (...\node_modules\@storybook\core-common\dist\cjs\presets.js:178:16)
> ERR!     at getContent (...\node_modules\@storybook\core-common\dist\cjs\presets.js:194:10)
> ERR!     at loadPreset (...\node_modules\@storybook\core-common\dist\cjs\presets.js:203:20)
> ERR!     at ...\node_modules\@storybook\core-common\dist\cjs\presets.js:256:18
> ERR!     at Array.reduce (<anonymous>)
> ERR!     at loadPresets (...\node_modules\@storybook\core-common\dist\cjs\presets.js:255:18)
> ERR!     at loadPreset (...\node_modules\@storybook\core-common\dist\cjs\presets.js:225:81)
> ERR!     at ...\node_modules\@storybook\core-common\dist\cjs\presets.js:256:18
> ERR!     at Array.reduce (<anonymous>)
> ERR!     at loadPresets (...\node_modules\@storybook\core-common\dist\cjs\presets.js:255:18)
> ERR!     at getPresets (...\node_modules\@storybook\core-common\dist\cjs\presets.js:313:23)
> ERR!     at loadAllPresets (...\node_modules\@storybook\core-common\dist\cjs\presets.js:365:10)
> ERR!     at buildDev (...\node_modules\@storybook\core-server\dist\cjs\build-dev.js:209:50)
> ERR!     at process.processTicksAndRejections (node:internal/process/task_queues:95:5) {
> ERR!   code: 'ERR_INVALID_ARG_TYPE'
> ERR! }
> ```
> 
> I am using react, next.js, tailwind, and typescript in this project. I had some issues before too but I think I got to work around those by adding some addons to storybook (some of these might be the reason that is causing these errors?) Here are my addons
> 
> ```
> addons: [
>     '@storybook/addon-links',
>     "@storybook/addon-essentials",
>     "@storybook/addon-interactions",
>     "@storybook/addon-a11y",
>     "storybook-dark-mode",
>     {
>       name: "@storybook/addon-postcss",
>       options: {
>         postcssLoaderOptions: {
>           implementation: require('postcss'),
>         },
>       },
>     },
>     "storybook-tailwind-dark-mode",
>     "storybook-addon-next",
>     "storybook-addon-next-router",
>     "storybook-addon-mock/register",
> ]
> ```
> 
> (with the post CSS it was having trouble with posttCSS8 so I remove this package and got the postCSS7)
> 
> I am not sure what makes storybook work the previous day but not work the next day, without changing anything.
> 
> I'd appreciate it if someone can point me where I could go from here, thanks!

I think the problem was with the postCSS and I was able to work around this by running `yarn add -D postcss@^8` in the terminal

But I got a different error 
```
ModuleParseError: Module parse failed: The keyword 'public' is reserved (36:16)
File was processed with these loaders:
 * ./node_modules/babel-loader/lib/index.js
You may need an additional loader to handle the result of these loaders.
|   var _super = _createSuper(***);
|
>   function FunctionName(1 , 2 = "", 3 = "", 4 = null, 5 = []) {
|     _classNAME(this, *);
|
    at handleParseError (...\node_modules\webpack\lib\NormalModule.js:976:19)
    at ...\node_modules\webpack\lib\NormalModule.js:1095:5
    at processResult (...\node_modules\webpack\lib\NormalModule.js:800:11)
    at ...\node_modules\webpack\lib\NormalModule.js:860:5
    at ...\node_modules\loader-runner\lib\LoaderRunner.js:407:3
    at iterateNormalLoaders (...\node_modules\loader-runner\lib\LoaderRunner.js:233:10)
    at iterateNormalLoaders (...\node_modules\loader-runner\lib\LoaderRunner.js:240:10)
    at ...\node_modules\loader-runner\lib\LoaderRunner.js:255:3
    at context.callback (...\node_modules\loader-runner\lib\LoaderRunner.js:124:13)
    at ...\node_modules\babel-loader\lib\index.js:59:71
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
ModuleParseError: Module parse failed: The keyword 'public' is reserved (32:70)
File was processed with these loaders:
 * ./node_modules/babel-loader/lib/index.js
You may need an additional loader to handle the result of these loaders.
|
| // Each node
> export var varName = /*#__PURE__*/_createClass(function varName(public something1, public someting2 = 0) {
|   _className(this, something);
| });
    at handleParseError (...node_modules\webpack\lib\NormalModule.js:976:19)
    at ...\node_modules\webpack\lib\NormalModule.js:1095:5
    at processResult (...\node_modules\webpack\lib\NormalModule.js:800:11)
    at ...\node_modules\webpack\lib\NormalModule.js:860:5
    at ...\node_modules\loader-runner\lib\LoaderRunner.js:407:3
    at iterateNormalLoaders (...\node_modules\loader-runner\lib\LoaderRunner.js:233:10)
    at iterateNormalLoaders (...\node_modules\loader-runner\lib\LoaderRunner.js:240:10)
    at ...\node_modules\loader-runner\lib\LoaderRunner.js:255:3
    at context.callback (...\node_modules\loader-runner\lib\LoaderRunner.js:124:13)
    at ...\node_modules\babel-loader\lib\index.js:59:71
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
ModuleParseError: Module parse failed: The keyword 'public' is reserved (34:23)
File was processed with these loaders:
 * ./node_modules/babel-loader/lib/index.js
You may need an additional loader to handle the result of these loaders.
|   var _super = _createSuper(something);
|
>   function functionName(arg1, arg2) {
|     _classVarName(this, something);
|
    at handleParseError (...\node_modules\webpack\lib\NormalModule.js:976:19)
    at ...\node_modules\webpack\lib\NormalModule.js:1095:5
    at processResult (...\node_modules\webpack\lib\NormalModule.js:800:11)
    at ...\node_modules\webpack\lib\NormalModule.js:860:5
    at ...\node_modules\loader-runner\lib\LoaderRunner.js:407:3
    at iterateNormalLoaders (...\node_modules\loader-runner\lib\LoaderRunner.js:233:10)
    at iterateNormalLoaders (...\node_modules\loader-runner\lib\LoaderRunner.js:240:10)
    at ...\node_modules\loader-runner\lib\LoaderRunner.js:255:3
    at context.callback (...\node_modules\loader-runner\lib\LoaderRunner.js:124:13)
    at ...\node_modules\babel-loader\lib\index.js:59:71
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
```

**Reply by mnlfischer** [reply]
I got this error when I forgot to install the @storybook/addon-postcss addon.

**Comment by m4th3w561** [comment, marked as answer]
Just an update.

I figured out how to go around the errors even if you have added the `@storybook/addon-postcss` and `postcss@^8` in react projects that use Tailwind/Storybook/Next.js 

Here is the flow I did to go around this error:
In your project folder, open terminal and run the following commands (If storybook is already installed you can skip step 1)
1. `npx storybook init`
2. `yarn add tailwindcss@yarn:@tailwindcss/postcss7-compat postcss@^7 autoprefixer@^9`
3. `yarn add -D @storybook/addon-postcss`
4. `yarn add -D postcss@^8`

In your .storybook/main.js add this block of code in the addon list:
```
addon: [
// other addons
    {
      name: "@storybook/addon-postcss",
      options: {
        postcssLoaderOptions: {
          implementation: require('postcss'),
        },
      },
    },
// other addons
  ]
```
